lib/types.nix

```diff
@@ -525,6 +529,31 @@ rec {
       modules = toList modules;
     };
 
+    # The type of a type!
+    optionType = mkOptionType {
+      name = "optionType";
+      description = "optionType";
+      check = value: value ? _type && value._type == "option-type";
+      merge = loc: defs:
+        let
+          # Prepares the type definitions for mergeOptionDecls, which
+          # annotates submodules types with file locations
+          optionModules = map ({ value, file }:
+            {
+              _file = file;
+              # There's no way to merge types directly from the module system,
+              # but we can cheat a bit by just declaring an option with the type
+              options = lib.mkOption {
+                type = value;
+              };
+            }
+          ) defs;
+          # Merges all the types into a single one, including submodule merging.
+          # This also propagates file information to all submodules
+          mergedOption = fixupOptionType loc (mergeOptionDecls loc optionModules);
```

Comment:
Do we need `fixupOptionType`?
I don't think this type will be used on (legacy) option sets.

Reply:
`fixupOptionType` actually seems to do a bit more than just what it says on the tin. Specifically it also does the `getSubModules`/`substSubModules` business, which makes sure that all submodules provided by all option declarations get merged together into the same type.

So I think it's better to just reuse `fixupOptionType` for now. In the future when the legacy stuff is removed we should probably rename this function for more clarity (or integrate it into `mergeOptionDecls`). I think that's outside of the scope of this PR though.